File: DESC_Python/src/update_data_spec.py
import os
import numpy as np
from scipy.io import loadmat, savemat
import multitaper
from round_half_up import round_half_up


def _matstruct_to_dict(obj):
    """Recursively convert MATLAB structs to Python dicts."""
    if isinstance(obj, np.ndarray) and obj.dtype.names:
        return {name: _matstruct_to_dict(obj[name]) for name in obj.dtype.names}
    if hasattr(obj, "_fieldnames"):
        return {fn: _matstruct_to_dict(getattr(obj, fn)) for fn in obj._fieldnames}
    if isinstance(obj, np.ndarray):
        return np.array(obj)
    return obj


def _multitaper_psd(signal, fs, nfft):
    mt = multitaper.MTSpec(signal, nw=2.5, kspec=3, dt=1/fs, nfft=nfft)
    spec = mt.spec.squeeze()[:nfft // 2 + 1]
    freq = mt.freq.squeeze()[:nfft // 2 + 1]
    if np.isnan(spec[0]):
        spec = np.zeros(spec.shape)
    return np.sqrt(spec), freq


def update_data_spec(evfile, disp_flag=1):
    """
    Compute spectra
    """
    mat = loadmat(evfile, squeeze_me=True, struct_as_record=False)
    if "data" not in mat:
        return
    data = _matstruct_to_dict(mat["data"])
    if data.get("numts", 0) == 0:
        return

    if disp_flag:
        print(f"Working on {evfile}")

    Nfft = 1024
    Nfftc = 4096

    numts = int(data["numts"])
    samprate_arr = np.asarray(data["samprate"]).flatten()

    def zmat(shape):
        return np.zeros(shape)

    data["pseis"] = np.empty(numts, dtype=object)
    data["pnoise"] = np.empty(numts, dtype=object)
    data["sseis"] = np.empty(numts, dtype=object)
    data["snoise"] = np.empty(numts, dtype=object)
    data["cseis"] = np.empty(numts, dtype=object)
    data["cnoise"] = np.empty(numts, dtype=object)
    data["pspec"] = zmat((numts, Nfft // 2 + 1))
    data["sspec"] = zmat((numts, Nfft // 2 + 1))
    data["cspec"] = zmat((numts, Nfftc // 2 + 1))
    data["pnspec"] = zmat((numts, Nfft // 2 + 1))
    data["snspec"] = zmat((numts, Nfft // 2 + 1))
    data["cnspec"] = zmat((numts, Nfftc // 2 + 1))
    data["pfreq"] = zmat((numts, Nfft // 2 + 1))
    data["sfreq"] = zmat((numts, Nfft // 2 + 1))
    data["cfreq"] = zmat((numts, Nfftc // 2 + 1))

    data["np1"] = zmat(numts)
    data["tp1"] = zmat(numts)
    data["ns1"] = zmat(numts)
    data["ts1"] = zmat(numts)
    data["codawin"] = zmat(numts)
    data["tswin"] = zmat(numts)
    data["tpwin"] = zmat(numts)
    data["pwin_mode"] = np.empty(numts, dtype=object)
    data["pnwin"] = zmat(numts)
    data["snwin"] = zmat(numts)
    data["cnwin"] = zmat(numts)
    data["pstn"] = zmat(numts)
    data["sstn"] = zmat(numts)

    samprate = samprate_arr[0]
    N_freq_samples = 1000 if samprate > 100 else 500
    data["flog"] = np.logspace(np.log10(0.5), np.log10(samprate * 0.45), N_freq_samples)

    # local-time adjustment
    data["stime"] += 8 * 3600

    for ns in range(numts):
        ppick = data["pick1"][ns]
        spick = data["pick2"][ns]
        time1 = data["stime"][ns]
        fs = data["samprate"][ns]
        npts = int(data["npts"][ns])

        if ppick > 0:
            data["np1"][ns] = int(np.floor((ppick + data["qtime"] - time1) * fs))
            data["tp1"][ns] = data["np1"][ns] / fs
        else:
            data["np1"][ns] = -12345
            data["tp1"][ns] = -12345

        if spick > 0:
            data["ns1"][ns] = int(np.floor((spick + data["qtime"] - time1) * fs))
            data["ts1"][ns] = data["ns1"][ns] / fs
        else:
            data["ns1"][ns] = -12345
            data["ts1"][ns] = -12345

        data["codawin"][ns] = 10.0
        qmag = data.get("qmag1", data.get("mb", 0.0))
        if qmag >= 4.0:
            data["tswin"][ns] = 10.0
        elif qmag >= 2.5:
            data["tswin"][ns] = 5.0
        else:
            data["tswin"][ns] = 3.0
        data["tpwin"][ns] = 2.0

        if ppick < 0:
            data["tpwin"][ns] = -12345
            data["pwin_mode"][ns] = ""
        elif spick < 0:
            data["pwin_mode"][ns] = "fixed"
        else:
            data["pwin_mode"][ns] = "psdiff"
            data["tpwin"][ns] = min(data["tpwin"][ns], spick - ppick - 0.1)

        # P-wave spectra
        if data["np1"][ns] > 0 and data["tpwin"][ns] >= 0.5:
            np1 = int(data["np1"][ns] - round_half_up(0.1 * fs))
            if np1 < 1:
                np1 = 1
            nwin = int(round_half_up(data["tpwin"][ns] * fs))
            nend = min(np1 + nwin - 1, npts)
            nwin = nend - np1 + 1
            data["pnwin"][ns] = nwin
            seis = np.asarray(data["seis"][ns]).flatten()
            pseg = seis[np1 - 1:nend]
            if np1 - nwin < 1:
                noise = np.zeros(nwin)
                noise[: np1 - 1] = seis[: np1 - 1]
            else:
                noise = seis[np1 - nwin - 1: np1 - 1]
            data["pseis"][ns] = pseg
            data["pnoise"][ns] = noise
            if nwin >= 50:
                spec, f = _multitaper_psd(pseg, fs, Nfft)
                nspec, _ = _multitaper_psd(noise, fs, Nfft)
                omega = 2 * np.pi * f
                nfg = f > 0
                data["pspec"][ns, 0] = spec[0]
                data["pnspec"][ns, 0] = nspec[0]
                data["pspec"][ns, nfg] = spec[nfg] / omega[nfg]
                data["pnspec"][ns, nfg] = nspec[nfg] / omega[nfg]
                data["pfreq"][ns, nfg] = f[nfg]
                specres = np.interp(data["flog"], f[nfg], data["pspec"][ns, nfg])
                data.setdefault("pspecres", np.zeros((numts, len(data["flog"]))))
                data["pspecres"][ns, : len(specres)] = specres

        # S-wave spectra
        if data["ns1"][ns] > 0:
            ns1 = max(int(data["ns1"][ns]), 1)
            nwin = int(round_half_up(data["tswin"][ns] * fs))
            nend = min(ns1 + nwin - 1, npts)
            nwin = nend - ns1 + 1
            data["snwin"][ns] = nwin
            seis = np.asarray(data["seis"][ns]).flatten()
            sseg = seis[ns1 - 1: ns1 - 1 + nwin]
            if ns1 - nwin < 1:
                snoise = np.zeros(nwin)
                snoise[: ns1 - 1] = seis[: ns1 - 1]
            else:
                snoise = seis[ns1 - nwin - 1: ns1 - 1]
            data["sseis"][ns] = sseg
            data["snoise"][ns] = snoise
            if nwin >= 50:
                spec, f = _multitaper_psd(sseg, fs, Nfft)
                nspec, _ = _multitaper_psd(snoise, fs, Nfft)
                omega = 2 * np.pi * f
                nfg = f > 0
                data["sspec"][ns, 0] = spec[0]
                data["snspec"][ns, 0] = nspec[0]
                data["sspec"][ns, nfg] = spec[nfg] / omega[nfg]
                data["snspec"][ns, nfg] = nspec[nfg] / omega[nfg]
                data["sfreq"][ns, nfg] = f[nfg]
                specres = np.interp(data["flog"], f[nfg], data["sspec"][ns, nfg])
                data.setdefault("sspecres", np.zeros((numts, len(data["flog"]))))
                data["sspecres"][ns, : len(specres)] = specres

            # Coda (approximate)
            ns1_c = max(ns1 - int(round_half_up(0.1 * fs)), 1)
            nwin_c = int(round_half_up(data["codawin"][ns] * fs))
            nend_c = min(ns1_c + nwin_c - 1, npts)
            nwin_c = nend_c - ns1_c + 1
            data["cnwin"][ns] = nwin_c
            seis = np.asarray(data["seis"][ns]).flatten()
            cseg = seis[ns1_c - 1: nend_c]
            if ns1_c - nwin_c < 1:
                cnoise = np.zeros(nwin_c)
                cnoise[: ns1_c - 1] = seis[: ns1_c - 1]
            else:
                cnoise = seis[ns1_c - nwin_c - 1: ns1_c - 1]
            spec, f = _multitaper_psd(cseg, fs, Nfftc)
            nspec, _ = _multitaper_psd(cnoise, fs, Nfftc)
            omega = 2 * np.pi * f
            nfg = f > 0
            data["cseis"][ns] = cseg
            data["cnoise"][ns] = cnoise
            data["cspec"][ns, 0] = spec[0]
            data["cnspec"][ns, 0] = nspec[0]
            data["cspec"][ns, nfg] = spec[nfg] / omega[nfg]
            data["cnspec"][ns, nfg] = nspec[nfg] / omega[nfg]
            data["cfreq"][ns, nfg] = f[nfg]
            specres = np.interp(data["flog"], f[nfg], data["cspec"][ns, nfg])
            data.setdefault("cspecres", np.zeros((numts, len(data["flog"]))))
            data["cspecres"][ns, : len(specres)] = specres

    # Ensure unfilled object array elements are set to empty arrays
    for ns in range(numts):
        if data["pseis"][ns] is None:
            data["pseis"][ns] = np.array([])
        if data["pnoise"][ns] is None:
            data["pnoise"][ns] = np.array([])
        if data["sseis"][ns] is None:
            data["sseis"][ns] = np.array([])
        if data["snoise"][ns] is None:
            data["snoise"][ns] = np.array([])
        if data["cseis"][ns] is None:
            data["cseis"][ns] = np.array([])
        if data["cnoise"][ns] is None:
            data["cnoise"][ns] = np.array([])

    savemat(evfile, {"data": data}, do_compression=True)

File: DESC_MATLAB/src/update_data_spec.m
% this is to recompute spectrum for data files.
function [] = update_data_spec(evfile,disp_flag)

if(nargin == 1)
    disp_flag = 1; % when disp_flag=1, the function will show matfile path
end

load(evfile);
if (~exist('data')); return; end

if (data.numts == 0)
    return;
end

if(disp_flag == 1)
    disp(strcat('Working on',evfile));
end
Nfft = 1024;
Nfftc = 4096;

data.pspec = zeros(data.numts, Nfft/2+1);
data.sspec = zeros(data.numts, Nfft/2+1);
data.cspec = zeros(data.numts, Nfftc/2+1);
data.pnspec = zeros(data.numts, Nfft/2+1);
data.snspec = zeros(data.numts, Nfft/2+1);
data.cnspec = zeros(data.numts, Nfft/2+1);
data.pfreq = zeros(data.numts, Nfft/2+1);
data.sfreq = zeros(data.numts, Nfft/2+1);
data.cfreq = zeros(data.numts, Nfft/2+1);

data.np1 = zeros(1, data.numts);
data.tp1 = zeros(1, data.numts);
data.ns1 = zeros(1, data.numts);
data.ts1 = zeros(1, data.numts);
data.codawin = zeros(1, data.numts);
data.tswin = zeros(1, data.numts);
data.tpwin = zeros(1, data.numts);
data.pwin_mode = cell(1, data.numts);
data.pnwin = zeros(1, data.numts);
data.snwin = zeros(1, data.numts);
data.cnwin = zeros(1, data.numts);
data.pstn = zeros(1, data.numts);
data.sstn = zeros(1, data.numts);

samprate = data.samprate(1);
N_freq_samples = 500;
if (round(samprate) > 100)
    N_freq_samples = 1000;
end
data.flog = logspace(log10(0.5), log10(samprate * 0.45), N_freq_samples);

data.pspec = zeros(data.numts, Nfft / 2 + 1);
data.sspec = zeros(data.numts, Nfft / 2 + 1);
data.cspec = zeros(data.numts, Nfftc / 2 + 1);
data.pnspec = zeros(data.numts, Nfft / 2 + 1);
data.snspec = zeros(data.numts, Nfft / 2 + 1);
data.cnspec = zeros(data.numts, Nfftc / 2 + 1);
data.pfreq = zeros(data.numts, Nfft / 2 + 1);
data.sfreq = zeros(data.numts, Nfft / 2 + 1);
data.cfreq = zeros(data.numts, Nfftc / 2 + 1);

%%% Only for local-time cases, adjust UTC to Beijing Time
data.stime = data.stime + 8*3600;

for ns = 1:data.numts
    ppick = data.pick1(ns);
    spick = data.pick2(ns);

    time1 = data.stime(ns);
    samprate = data.samprate(ns);

    if (ppick > 0)
        data.np1(ns) = floor((ppick + data.qtime - time1)*samprate);
        data.tp1(ns) = data.np1(ns) / samprate;
    else
        data.np1(ns) = -12345;
        data.tp1(ns) = -12345;
    end


    if (spick > 0)
        data.ns1(ns) = floor((spick + data.qtime - time1)*samprate);
        data.ts1(ns) = data.ns1(ns)/samprate;
    else
        data.ns1(ns) = -12345;
        data.ts1(ns) = -12345;
    end

    % Define windows for spectra calculation
    data.codawin(ns) = 10;
    try
        qmag = data.qmag1;
    catch
        qmag = data.mb;
    end
    %%% Define your own S window lengths for different magnitudes
    if (qmag >= 4.0)
        data.tswin(ns) = 10.0;
    elseif (qmag >= 2.5)
        data.tswin(ns) = 5.0;
    else
        data.tswin(ns) = 3.0;
    end
    %%% Define your own P window lengths according to your waveforms
    data.tpwin(ns) = 2.0;
    % if (qmag >= 4.0)
    %     data.tpwin(ns) = 2.0;
    % elseif (qmag >= 2.5)
    %     data.tpwin(ns) = 2.0;
    % else
    %     data.tpwin(ns) = 2.0;
    % end
    if (ppick < 0)
        data.tpwin(ns) = -12345;
        data.pwin_mode{ns} = '';
    elseif (spick < 0)
        data.pwin_mode{ns} = 'fixed';
    else
        data.pwin_mode{ns} = 'psdiff';
    end
    if(strcmp(data.pwin_mode{ns},'psdiff'))     % Use S-P as P window
        data.tpwin(ns) = min(data.tpwin(ns), spick - ppick - 0.1);
    end


    %########### now will compute P and S signal to noise ratio with 10-40 bandpass
    data.pstn(ns) = 0;
    data.sstn(ns) = 0;

    if (data.np1(ns) > 0 && data.tpwin(ns) >= 0.5)
        %### This is for P wave.
        np1 = data.np1(ns) - round(0.1 * data.samprate(ns)); %allow 0.1 second before
        if (np1 < 1)
            np1 = 1;
        end
        nwin = round(data.tpwin(ns) * data.samprate(ns));
        nend = np1 + nwin - 1;
        if (nend > data.npts(ns))
            nend = data.npts(ns);
            nwin = data.npts(ns) - np1 + 1;
        end
        data.pnwin(ns) = nwin;
        data.pseis(ns,1:nwin) = data.seis(ns,np1:nend);
        if (np1 - nwin < 1)
            data.pnoise(ns,1:nwin) = zeros(1,nwin);
            data.pnoise(ns,1:np1-1) = data.seis(ns,1:np1-1);
        else
            data.pnoise(ns,1:nwin) = data.seis(ns,np1-nwin:np1-1);
        end
        if (nwin >= 50)
            nx = Nfft;
            [spec,~] = pmtm(data.pseis(ns,1:nwin),3,nx,data.samprate(ns));
            [nspec,f] = pmtm(data.pnoise(ns,1:nwin),3,nx,data.samprate(ns));
            omega = 2 * pi * f;
            nfg = find(f > 0);
            spec = sqrt(spec);
            nspec = sqrt(nspec);
            data.pspec(ns,1) = spec(1);
            data.pnspec(ns,1) = nspec(1);
            data.pspec(ns,nfg) = spec(nfg) ./ omega(nfg);
            data.pnspec(ns,nfg) = nspec(nfg) ./ omega(nfg);
            data.pfreq(ns,nfg) = f(nfg);
            % nband = find(f(nfg)>=4 & f(nfg)<=10);
            % data.pstn(ns) = mean(log10(data.pspec(ns,nband))-log10(data.pnspec(ns,nband)));
            specres = interp1(f(nfg),data.pspec(ns,nfg),data.flog,'linear','extrap');
            data.pspecres(ns,1:length(data.flog)) = specres;
        end
    end

    if (data.ns1(ns) > 0)
        %#### this is for S-wave.
        ns1 = data.ns1(ns);
        if (ns1 < 1)
            ns1 = 1;
        end
        nwin = round(data.tswin(ns)*data.samprate(ns));
        nend = ns1+nwin-1;
        if (nend > data.npts(ns))
            nend = data.npts(ns);
            nwin = data.npts(ns)-ns1+1;
        end
        data.snwin(ns) = nwin;
        if (nwin >= 50)

            data.sseis(ns,1:nwin) = data.seis(ns,ns1:ns1+nwin-1);
            if (ns1-nwin<1)
                data.snoise(ns,1:nwin) = zeros(1,nwin);
                data.snoise(ns,1:ns1-1) = data.seis(ns,1:ns1-1);
            else
                data.snoise(ns,1:nwin) = data.seis(ns,ns1-nwin:ns1-1);
            end
            nx = Nfft;

            [spec,~] = pmtm(data.sseis(ns,1:nwin),3,nx,data.samprate(ns));
            [nspec,f] = pmtm(data.snoise(ns,1:nwin),3,nx,data.samprate(ns));
            omega = 2*pi*f;
            nfg = find(f>0);
            spec = sqrt(spec);
            nspec = sqrt(nspec);
            data.sspec(ns,1) = spec(1);
            data.snspec(ns,1) = nspec(1);
            data.sspec(ns,nfg) = spec(nfg)./omega(nfg);
            data.snspec(ns,nfg) = nspec(nfg)./omega(nfg);
            data.sfreq(ns,nfg) = f(nfg);
            % nband = find(f(nfg)>=4 & f(nfg)<=10);
            % data.sstn(ns) = mean(log10(data.sspec(ns,nband))-log10(data.snspec(ns,nband)));
            specres = interp1(f(nfg),data.sspec(ns,nfg),data.flog,'linear','extrap');
            data.sspecres(ns,1:length(data.flog)) = specres;
        end

        % this is for coda wave (Don't use yet. Not ready).
        ns1 = data.ns1(ns)-round(0.1*data.samprate(ns)); % allow 0.1 second before
        if (ns1<1) ns1 = 1; end
        nwin = round(data.codawin(ns)*data.samprate(ns));
        nend = ns1+nwin-1;
        if (nend > data.npts(ns)) nend = data.npts(ns); nwin = data.npts(ns)-ns1+1; end
        data.cnwin(ns) = nwin;
        data.cseis(ns,1:nwin) = data.seis(ns,ns1:nend);
        nxx1 = ns1-nwin;
        nwin1 = nwin;

        %if (ns1-nwin<1)
        %    nxx1 = 5;
        %    nwin1 = ns1-nxx1;
        %end
        if (ns1-nwin<1)
            data.cnoise(ns,1:nwin) = zeros(1,nwin);
            data.cnoise(ns,1:ns1-1) = data.seis(ns,1:ns1-1);
        else
            data.cnoise(ns,1:nwin) = data.seis(ns,ns1-nwin:ns1-1);
        end
        %data.cnoise(ns,1:nwin1) = data.seis(ns,nxx1:ns1-1);
        %nx = 2^(nextpow2(nwin));
        nx = Nfftc;
        [spec,~] = pmtm(data.cseis(ns,1:nwin),3,nx,data.samprate(ns));
        [nspec,f] = pmtm(data.cnoise(ns,1:nwin1),3,nx,data.samprate(ns));
        %factor = nx*samprate(iw)*2*pi;
        omega = 2*pi*f;
        nfg = find(f>0);
        spec = sqrt(spec);
        nspec = sqrt(nspec);
        data.pspec(ns,1) = spec(1);
        data.pnspec(ns,1) = nspec(1);
        data.cspec(ns,nfg) = spec(nfg)./omega(nfg);
        data.cnspec(ns,nfg) = nspec(nfg)./omega(nfg);
        data.cfreq(ns,nfg) = f(nfg);
        specres = interp1(f(nfg),data.cspec(ns,nfg),data.flog,'linear','extrap');
        data.cspecres(ns,1:length(data.flog)) = specres;
    end

end

try
    data.pseis;
catch
    data.pseis = [];
    data.pnoise = [];
    data.pspecres = [];
end
try
    data.sseis;
catch
    data.sseis = [];
    data.snoise = [];
    data.sspecres = [];
end
try
    data.cseis;
catch
    data.cseis = [];
    data.cnoise = [];
    data.cspecres = [];
end
numTraceAvail = size(data.pseis, 1);
trace_diff = data.numts - numTraceAvail;

data.pseis = [data.pseis; zeros(trace_diff, size(data.pseis, 2))];
data.pnoise = [data.pnoise; zeros(trace_diff, size(data.pnoise, 2))];
data.pspecres = [data.pspecres; zeros(trace_diff, size(data.pspecres, 2))];

data.sseis = [data.sseis; zeros(trace_diff, size(data.sseis, 2))];
data.snoise = [data.snoise; zeros(trace_diff, size(data.snoise, 2))];
data.sspecres = [data.sspecres; zeros(trace_diff, size(data.sspecres, 2))];
data.cseis = [data.cseis; zeros(trace_diff, size(data.cseis, 2))];
data.cnoise = [data.cnoise; zeros(trace_diff, size(data.cnoise, 2))];
data.cspecres = [data.cspecres; zeros(trace_diff, size(data.cspecres, 2))];

save(evfile,'data','-v7.3');

